### PR TITLE
rofl threading rework and cleanup

### PR DESCRIPTION
**Pull request by toanju**
This PR enables the cthread class to be used by multiple instances as an event loop. All previous existing threads are collapsed into a single thread. Therefore callbacks can now be passed to cthread to cope e.g. with multiple wakeup events.

### CODE REVIEW

File: src/rofl/common/cthread.hpp [hunk 1]
```diff
@@ -219,24 +215,19 @@ class cthread {
...
+  int epfd;      // worker thread epoll fd
+  int wake_handle;
 
-  crwlock tlock; // thread lock
+  crwlock tlock;       // thread lock
+  crwlock events_lock; // events lock
 
   std::map<int, uint32_t> fds;     // set of registered file descriptors
   std::set<ctimer> ordered_timers; // ordered set of timers
+  int max_fds; // XXX(toanju): currently fixed to 1024, but should be read
```
Comment (DFritzsche) [inline src/rofl/common/cthread.hpp]: why not make a const?

File: src/rofl/common/ctimer.hpp [hunk 2]
```diff
@@ -92,6 +86,8 @@ class ctimer {
    */
   const ctimespec &get_tspec() const { return tspec; };
 
+  cthread_timeout_event *get_callback() { return e; }
```
Comment (DFritzsche) [inline src/rofl/common/ctimer.hpp]: make get function + return const?

File: test/rofl/common/crofbase/crofbasetest.cpp [hunk 3]
```diff
@@ -29,19 +29,15 @@ void crofbasetest::test() {
     ts.tv_sec = 1;
     ts.tv_nsec = 0;
     pselect(0, NULL, NULL, NULL, &ts, NULL);
-    std::cerr << "#";
+    LOG(INFO) << "#";
   }
-  std::cerr << std::endl;
+  LOG(INFO) << std::endl;
```
Comment (DFritzsche) [inline test/rofl/common/crofbase/crofbasetest.cpp]: excuse me?

File: test/rofl/common/crofbase/crofbasetest.cpp [hunk 4]
```diff
@@ -83,21 +79,21 @@ void cdatapath::test_start() {
 }
 
 void ccontroller::handle_dpt_open(rofl::crofdpt &dpt) {
-  std::cerr << ">>> XXX dpt connected: " << std::endl;
+  LOG(INFO) << ">>> XXX dpt connected: " << std::endl;
```
Comment (DFritzsche) [inline test/rofl/common/crofbase/crofbasetest.cpp]: should that XXX log the dpid? more below

File: test/rofl/common/crofchan/crofchantest.hpp [hunk 5]
```diff
@@ -143,118 +136,119 @@ class crofchantest : public CppUnit::TestFixture,
...
+  void congestion_occurred_indication(rofl::crofchan &chan,
+                                      rofl::crofconn &conn) override;
 
-  virtual void handle_transaction_timeout(rofl::crofchan &chan,
-                                          rofl::crofconn &conn, uint32_t xid,
-                                          uint8_t type, uint16_t sub_type = 0) {
+  void handle_transaction_timeout(rofl::crofchan &chan, rofl::crofconn &conn,
+                                  uint32_t xid, uint8_t type,
+                                  uint16_t sub_type = 0) override {
     std::cerr << "crofchan::handle_transaction_timeout" << std::endl;
```
Comment (DFritzsche) [inline test/rofl/common/crofchan/crofchantest.hpp]: maybe use GLOG here as well

File: examples/ethswctld/cfibentry.cc [hunk 6]
```diff
@@ -24,7 +24,8 @@ cfibentry::~cfibentry() {
   std::cerr << "[cfibentry] deleted" << std::endl << *this;
```
Comment (DFritzsche) [inline examples/ethswctld/cfibentry.cc]: use GLOG? more below

File: examples/ethswctld/cetherswitch.cc [hunk 7]
```diff
@@ -82,13 +82,14 @@ cetherswitch::cetherswitch(
 
 cetherswitch::~cetherswitch() {}
 
-void cetherswitch::handle_timeout(cthread &thread, uint32_t timer_id) {
+void cetherswitch::handle_timeout(void *userdata) {
   try {
+    int timer_id = (long)userdata;
```
Comment (toanju) [inline examples/ethswctld/cetherswitch.cc]: weird but necessary. Looking at this right now I think it should be a more c++ style cast :)

File: src/rofl/common/crofbase.h [hunk 8]
```diff
@@ -1169,9 +1169,9 @@ class crofbase : public virtual rofl::cthread_env,
    * @param auxid control connection identifier
    * @param msg OpenFlow message instance
    */
-  virtual void
-  handle_get_config_reply(rofl::crofdpt &dpt, const rofl::cauxid &auxid,
```
Comment (toanju) [inline src/rofl/common/crofbase.h]: indeed, but I even shouldn't have done these drive by fixes here. I will add the rest in an additional PR.

File: src/rofl/common/cthread.cpp [hunk 9]
```diff
@@ -202,9 +363,44 @@ void cthread::drop_read_fd(int fd, bool exception) {
...
+      delete events_reg[fd];
+      events_reg[fd] = nullptr;
+    }
+  }
 }
 
-void cthread::add_write_fd(int fd, bool exception, bool edge_triggered) {
+void cthread::add_write_fd(int fd, cthread_write_event *cb, void *userdata,
+                           bool exception, bool edge_triggered) {
+
```
Comment (toanju) [inline src/rofl/common/cthread.cpp]: For the registration it should not matter since the event loop cannot deal with the structure yet. I will double check.

File: src/rofl/common/cthread.hpp (hunk 1 above)
Comment (toanju) [inline src/rofl/common/cthread.hpp]: sort of explained in the comment, but still could be set constant right now.

File: src/rofl/common/ctimer.hpp (hunk 2 above)
Comment (toanju) [inline src/rofl/common/ctimer.hpp]: can't be constant otherwise callbacks and their implementations should be constant as well which shouldn't be the case

File: test/rofl/common/crofbase/crofbasetest.cpp (hunk 4 above)
Comment (toanju) [inline test/rofl/common/crofbase/crofbasetest.cpp]: not sure what the original intention was, but it could be done like this, will add it in a separate PR.

File: examples/ethswctld/cfibentry.cc (hunk 6 above)
Comment (toanju) [inline examples/ethswctld/cfibentry.cc]: Right, but should be done in a separate PR.

File: src/rofl/common/ctimer.hpp (hunk 2 above)
Comment (DFritzsche) [inline src/rofl/common/ctimer.hpp]: can we then rename the function to make it clearer?

**Comment by akoepsel** [comment]
superseded by PR #102
